### PR TITLE
Implementation of the converter service

### PR DESCRIPTION
**Pull request by alexey-stankevich**

### CODE REVIEW

File: converter_service/src/main/java/com/kyripay/converter/service/ConversionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@Service
+public class ConversionServiceImpl implements ConversionService
+{
+
+  private final DocumentRepostiory repostiory;
+
+
+  public ConversionServiceImpl(DocumentRepostiory repostiory)
```
Comment (xpinjection) [inline converter_service/src/main/java/com/kyripay/converter/service/ConversionServiceImpl.java]: You could use AllArgsConstructor annotation to simplify injection.

File: converter_service/src/main/java/com/kyripay/converter/converters/Format.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import com.kyripay.converter.dto.FormatDetails;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+
+@AllArgsConstructor
+@Getter
+public enum Format
+{
+  IDENTITY("Identity", IdentityConverter.class, "String representation of initial payment"),
```
Comment (xpinjection) [inline converter_service/src/main/java/com/kyripay/converter/converters/Format.java]: There is too much coupling between format and converter implementation. It is better to return supported formats from each converter and then inject them all to service, where you could find appropriate on request.

File: converter_service/src/test/java/com/kyripay/converter/api/ConverterTest.java [hunk 3]
```diff
@@ -41,6 +50,9 @@ public void setUp() {
             .build();
     }
 
+    @MockBean
+    private DocumentRepostiory mockedDocumentRepository;
```
Comment (xpinjection) [inline converter_service/src/test/java/com/kyripay/converter/api/ConverterTest.java]: API level test should work on real DB, use testcontainers to start MongoDB.

File: converter_service/src/main/resources/application.yml [hunk 4]
```diff
@@ -0,0 +1,12 @@
+spring:
+  application:
+    name: converter
+  data:
+    mongodb:
+      database: converter
+
+eureka:
```
Comment (xpinjection) [inline converter_service/src/main/resources/application.yml]: If Eureka client is disabled why do you need it now?

File: converter_service/Dockerfile [hunk 5]
```diff
@@ -1,7 +1,11 @@
 FROM openjdk:11-jre-slim
 
-COPY build/libs/converter-0.0.1-SNAPSHOT.jar /usr/src/converter-0.0.1-SNAPSHOT.jar
+ARG VERSION
 
-CMD java -jar /usr/src/converter-0.0.1-SNAPSHOT.jar
+COPY converter-${VERSION}.jar /usr/src/converter-${VERSION}.jar
+
+ENV VERSION ${VERSION}
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/Dockerfile]: It was used in the CMD. Now I removed version from the deployed jar name.

File: converter_service/src/main/java/com/kyripay/converter/api/ConverterController.java [hunk 6]
```diff
@@ -1,60 +1,87 @@
...
   }
 
 
   @GetMapping("converters")
   @ApiOperation("Returns list of available formats")
-  List<Format> getFormats()
+  List<FormatDetails> getFormats()
   {
-    return Lists.newArrayList(Format.FORMAT_1, Format.FORMAT_2);
+    return Arrays.stream(Format.values()).map(Format::asFormatDetails).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/api/ConverterController.java]: Done

File: converter_service/src/main/java/com/kyripay/converter/api/ConverterController.java [hunk 7]
```diff
@@ -1,60 +1,87 @@
...
+  )
   Document getDocument(@NotBlank(message = "Document id must be provided") @PathVariable String id)
   {
-    return new Document(id, Format.FORMAT_1, "converted data".getBytes());
+    return conversionService.getDocument(id);
   }
 
 
   @Value
   @ApiModel(value = "Converter response", description = "Contains id of the document to be requested via '/api/v1/documents/{id}' endpoint")
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/api/ConverterController.java]: Done

File: converter_service/src/main/java/com/kyripay/converter/converters/SlowIdentityConverter.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.concurrent.TimeUnit;
+
+
+public class SlowIdentityConverter implements Converter
+{
+  @Override
+  public byte[] convert(Payment payment)
+  {
+    try {
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/converters/SlowIdentityConverter.java]: It was used to see the asynchrony with the naked eye :), changed to something closer to the real life.

Could you give more info about local application events?

File: converter_service/src/main/java/com/kyripay/converter/exceptions/DocumentNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package com.kyripay.converter.exceptions;
+
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+
+@ResponseStatus(HttpStatus.NOT_FOUND)
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/exceptions/DocumentNotFoundException.java]: done

File: converter_service/src/main/java/com/kyripay/converter/exceptions/DocumentNotFoundException.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+public class DocumentNotFoundException extends RuntimeException
+{
+  public DocumentNotFoundException(String message)
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/exceptions/DocumentNotFoundException.java]: done

File: converter_service/src/main/java/com/kyripay/converter/converters/Format.java (hunk 2 above)
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/converters/Format.java]: Implemantation changed

File: converter_service/src/test/java/com/kyripay/converter/api/ConverterTest.java (hunk 3 above)
Comment (alexey-stankevich) [reply converter_service/src/test/java/com/kyripay/converter/api/ConverterTest.java]: done

File: converter_service/src/main/java/com/kyripay/converter/converters/Converter.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package com.kyripay.converter.converters;
+
+import com.kyripay.converter.dto.FormatDetails;
+import com.kyripay.converter.dto.Payment;
+import org.springframework.stereotype.Component;
+
+
+@Component
```
Comment (MoonWorm) [inline converter_service/src/main/java/com/kyripay/converter/converters/Converter.java]: is it a good idea to have Spring IoC annotation on interface level?

Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/converters/Converter.java]: I forgot to remove it)

File: converter_service/src/main/java/com/kyripay/converter/converters/SlowIdentityConverter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (xpinjection) [reply converter_service/src/main/java/com/kyripay/converter/converters/SlowIdentityConverter.java]: Here is basic guide: https://www.baeldung.com/spring-events

[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/converters/SlowIdentityConverter.java]: Events implemented

File: converter_service/src/main/java/com/kyripay/converter/exceptions/WrongFormatException.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
+package com.kyripay.converter.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+
+@ResponseStatus(HttpStatus.BAD_REQUEST)
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/exceptions/WrongFormatException.java]: done

File: converter_service/src/main/java/com/kyripay/converter/domain/PaymentDocument.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+
+import javax.validation.constraints.NotNull;
+
+
+@Document
+@Data
+@AllArgsConstructor
+public class PaymentDocument
```
[the remark this answers is not in the document]
Comment (alexey-stankevich) [reply converter_service/src/main/java/com/kyripay/converter/domain/PaymentDocument.java]: This class is removed, Document is used for both api and persistence. There's no create/update endpoint for it, so I this should be OK.

**Pull request merged**